Building a withdraw limit proposal system. Each owner can submit a withdraw limit for other accounts, to be signed by the owners.

Fils7 opened the pull request:
In this branch I've built a structure to keep track of an account withdraw timestamp and limit. 
The `submitAccountLimit` function lets each owner submit proposals for withdraw limits in others accounts, the proposals will be executed if the required signatures are met.
Like in a transaction, I used a mapping to store all proposals indexes.
The `signSettings` function, allows owners to sign on a specific proposal. This will call an internal function that executes all the logic of signing a proposal.
Made a `withdraw` function to be called by an owner. This owner can only withdraw up to the withdraw limit.

Fils7 committed:
Working on a proposal system to submit a withdraw limit to other owners

This feature allows an owner (#1) to submit withdraw limits for the other owners (#2, #3) or himself.
For this, we need to track several requests and "actions" within the wallet:

 .A structure to keep track of an account settings;
.Mapping of who signed each settings proposal.
.Mapping similar to txMap, to track all the proposals submitted to other owners.
.Another mapping to connect an Account to an Address, after reaching the requested signatures.

Made the submitAccountLimit() that gives an owner the power to submit/propose a withdraw limit for the address given as parameter. Will store the proposer,
the last withdraw timestamp of the account, how many approvals were made, the new withdraw limit and to, the account passed in the function.

Started the signAccountLImit() function that will sign a proposal for the accounts.
Working on the withdraw function.

Fils7 committed:
Building a proposal system for withdraw account limits.

.Modified the Account structure from address _to, to address _targetAccount. And added a boolean to see if the withdraw limit was executed.

.Separated the sign account limit function to an internal function that executes the logic.
(Can only execute if the required signatures are met & the proposal was not executed yet.)

.Made all the events for each function.

brunoffranca commented on contracts/Spades.sol:
```diff
@@ -113,65 +151,135 @@ contract Spades {
+    function getTransaction(uint txIndex) public view returns (address targetAccount, uint amount, uint confirmations, address signature, bool executed) {
     
-        Transaction memory transaction = txMap[txIndex];
+        Transaction storage transaction = txMap[txIndex];
```
```suggestion
        transaction = txMap[txIndex];
```
Won't this compile like this?

Fils7 replied:
This won't compile. You need to specify `storage`because memory is the default.

Fils7 replied:
Your suggestion doesn't work, it will assume `memory` as default. We need to specify `storage`.

brunoffranca commented on contracts/Spades.sol:
```diff
@@ -113,65 +151,135 @@ contract Spades {
+
+        require(_dailyWithdrawlimit >= 0, "Invalid dailyWithdrawlimit");
+        require(OwnersCheck[_account] = true, "Address does not match with any of the owners.");
+        uint accountTimestamp = accountInfo[_account].lastWithdrawalTimestamp;
```
There might not be an entry in that array. If I'm creating a limit for that owner for the first time, this will fail.

brunoffranca commented on contracts/Spades.sol:
```diff
@@ -113,65 +151,135 @@ contract Spades {
+   }
+
+
+   function signSettings(uint256 _proposedSettings) public ownerOnly {
+    signFunction(_proposedSettings);
+
+    emit signedSettings(msg.sender, _proposedSettings);
+
+}
+
+    ///@dev This signs the request for the setting, after reaching
+    // the required signatures will execute by itself. Can only be called within this contract.
+
+   function signFunction(uint _proposedSettings) internal {
+
+        require(whoSignedSettings[_proposedSettings][msg.sender] == false, "Owner already signed this settings");
+        Account storage account = submitedSettings[_proposedSettings];
+        account.approvals += 1;
+        whoSignedSettings[_proposedSettings][msg.sender] == true;
+
+        if (account.approvals >= requiredSignatures && !account.executed) {
+            accountInfo[account.targetAccount].dailyWithdrawlimit = account.dailyWithdrawlimit;
+            account.executed = true;
+            
+        }
+
+   }
```
Why have separate functions?

brunoffranca commented on contracts/Spades.sol:
```diff
@@ -113,65 +151,135 @@ contract Spades {
+        whoSignedSettings[_proposedSettings][msg.sender] == true;
+
+        if (account.approvals >= requiredSignatures && !account.executed) {
+            accountInfo[account.targetAccount].dailyWithdrawlimit = account.dailyWithdrawlimit;
```
This line is not doing anything.

Fils7 replied:
Could you elaborate please?

brunoffranca commented on contracts/Spades.sol:
```diff
@@ -42,43 +68,55 @@ contract Spades {
+
+
+    ///@dev Conects the account targetAccount an address
+    mapping (address => Account) public accountInfo;
```
It would be better to have a separate struct for this mapping, instead of using the Account struct. Account contains data that you don't need to keep long-term, and it's just wasteful. Also you want to be able to clean the txMap and submittedSettings maps when the transactions in there are executed.

Fils7 replied:
It is not possible to delete any storage data from the EVM once you write it.
I can try to search for a workaround maybe.

brunoffranca reviewed (changes requested):
There's a bunch of typos that you need to fix. You also need to improve the naming of the structs. Other than that, it looks decent.

Fils7 committed:
Correction and contract organization.

After the PR review I  made some adjustments to the contract:

.Organized the code with comment fields.
.Organized comment lines on all variables.
.Made a new struct `Settings`to keep track of account settings.
.Deleted unnecessary fields from struct `ProposedSettings` and struct `Transaction`.
.Made a mapping `whoSignedProposedSettings` to keep track of confirmations by the owners.
.Deleted all stored values after being executed to clear storage space.
.Changed the function `signSettings` to pass the values to struct `Settings`at the end of execution.
